### DISCUSSION TITLE
How to get all components for a specific entity?

### DISCUSSION THREAD

**Discussion by MrGVSV** [bevyengine/bevy#3332] [Q&A] [answered]
> Asking and answering this for other users' who might have the same question

Is there a way to print a list of all components on an entity for debugging purposes?

**Comment by MrGVSV** [comment]
Here's a helper function for collecting the `ComponentId` for an entity's archetype:

```rust
pub fn get_components_for_entity<'a>(
    entity: &Entity,
    archetypes: &'a Archetypes,
) -> Option<impl Iterator<Item = ComponentId> + 'a> {
    for archetype in archetypes.iter() {
        if archetype.entities().contains(entity) {
            return Some(archetype.components());
        }
    }
    None
}
```

And you can use it like this:

```rust
// The system to print the components
pub fn my_system(archetypes: &Archetypes, components: &Components, /** ... */) {

  // ...

  for comp_id in get_components_for_entity(&my_entity, archetypes).unwrap() {
    if let Some(comp_info) = components.get_info(comp_id) {
      println!("Component: {:?}", comp_info);
    }
  }
}
```

**Reply by OriginLive** [reply]
Doesn't work with bevy 0.9 :(

**Reply by kristoff3r** [reply]
> Doesn't work with bevy 0.9 :(

In bevy 0.9 you can replace 

```
if archetype.entities().contains(entity) {
```

With

```            
if archetype.entities().iter().any(|e| e.entity() == entity) {
```

**Comment by mrchantey** [comment]
Here's a modification that I use:

```rust
pub fn all_component_ids<'a>(
    world: &'a World,
    entity: Entity,
) -> impl Iterator<Item = ComponentId> + 'a {
    let components = world.components();
    for archetype in world.archetypes().iter() {
        if archetype.entities().iter().any(|e| e.entity() == entity) {
            return archetype.components();
        }
    }
    world.archetypes().empty().components()
}
pub fn all_component_infos<'a>(
    world: &'a World,
    entity: Entity,
) -> impl Iterator<Item = &'a ComponentInfo> + 'a {
    let components = world.components();
    all_component_ids(world, entity).map(|id| {
        components
            .get_info(id)
            .expect("Component id without info, this shouldnt happen..")
    })
}
```
Usage:

```rust
let names = all_component_infos(&app.world, entity)
        .map(|info| info.name())
        .collect::<Vec<_>>()
        .join("\n\n");
println!("{names}");
```

**Reply by MrGVSV** [reply]
Thanks for this! I think we can simplify even further though! I just posted a new answer [here](https://github.com/bevyengine/bevy/discussions/3332#discussioncomment-7418611) 😄

**Reply by mrchantey** [reply]
Awesome thanks for updating this!

**Comment by MrGVSV** [comment, marked as answer]
**Update:** This can be done a lot simpler now with just a `World` reference using [`World::inspect_entity`](https://docs.rs/bevy/latest/bevy/ecs/prelude/struct.World.html#method.inspect_entity).

```rust
fn my_system(world: &World) {
  // ...

  println!("{:#?}", world.inspect_entity(my_entity));
}
```

> I believe this method was added in v0.9, whereas my [original answer](https://github.com/bevyengine/bevy/discussions/3332#discussioncomment-1809893) was likely from v0.5.

**Reply by cstegel** [reply]
Thanks for the tip about [World::inspect_entity](https://docs.rs/bevy/latest/bevy/ecs/prelude/struct.World.html#method.inspect_entity)!  To build on this in case others find it (like I did), the debug printing of `ComponentInfo` is extremely verbose so it might be good to start with just the names. Also, it needs to be collected into a `Vec` to be printed:

```rust
fn my_system(world: &World) {
  // ...

  println!("{:#?}", world.inspect_entity(my_entity)
                         .map(|info| info.name())
                         .collect::<Vec<_>>());
}
```
